execstats: include local streams in the physical plan

asubiotto opened the pull request:
Fixes #60463 

Look at commits for individual details. Ignored errors caused this bug to go unnoticed, so the first commit adds a panic when the crdb test flag is specified. This would've caught the issue easily.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/60484)
<!-- Reviewable:end -->

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  Thanks!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60484#-:-MTH1gTI3mlc6Z9oGFO9:b-e7bejd)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
___
*[pkg/sql/instrumentation.go, line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60484#-MTHTCKKA7XE0J_mgqlM:-MTHTCKKA7XE0J_mgqlN:b-ddg0p9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d88099955fb58a938ebfcbc5cbc808824b405563/pkg/sql/instrumentation.go#L256)):*
> ```Go
> 			stmtStats.mu.Unlock()
> 		}
> 		txnStats.Accumulate(queryLevelStats)
> ```

I had an unrelated question here - do we not want to update txnStats when stmtStats is nil? I assume stmtStats is nil the first time we run a given statement fingerprint, which seems irrelevant if we care about txnstats?


<!-- Sent from Reviewable.io -->

asubiotto commented:
The check fails in some EXPLAIN ANALYZE edge cases (https://github.com/cockroachdb/cockroach/issues/60609). If the build is green, I'll move it along with the check disabled with modified output because this PR fixes + improves testing, but we should get to the bottom of the failure.

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/60484)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @RaduBerinde)

___
*[pkg/sql/instrumentation.go, line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/60484#-MTHTCKKA7XE0J_mgqlM:-MTebhjP-PP3l--F2vvf:b6a8tjr) ([raw file](https://github.com/cockroachdb/cockroach/blob/d88099955fb58a938ebfcbc5cbc808824b405563/pkg/sql/instrumentation.go#L256)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I had an unrelated question here - do we not want to update txnStats when stmtStats is nil? I assume stmtStats is nil the first time we run a given statement fingerprint, which seems irrelevant if we care about txnstats?
</blockquote></details>

No, `stmtStats` should never be nil during normal execution since they're recorded in `dispatchToExecutionEngine` (we should unify these two sites) before this call to `Finish`. I think the cases in which they can be nil are when `retErr != nil`.


<!-- Sent from Reviewable.io -->

Commit:
sql: panic with TraceAnalyzer errors in crdb_test builds

During normal operation, failure to calculate query stats should not return a
query error. However, this behavior meant we would not observe errors that
occur. This commit panics with any errors if the crdb_test flag is specified,
since an error implies unexpected state. This allows us to get visibility into
these types of errors.

Release note: None

Commit:
execstats: include local streams in the physical plan

Previously, only remote streams would be included in the physical plan. However
local streams are components too! For example, they might buffer or use disk.
Not including them would result in errors when stats were collected for plans
with local streams, since the stats would be present in the trace, but a
corresponding component couldn't be found.

Release note: None

RaduBerinde commented:
Sounds good, thanks!

asubiotto commented:
TFTR

bors r=RaduBerinde

asubiotto commented:
flake (https://github.com/cockroachdb/cockroach/issues/58955)

bors r=RaduBerinde
